TUTORIAL.md

```diff
@@ -0,0 +1,392 @@
+  [name, department_id]
+end
+
+def insert
+  sql = <<-SQL
+    INSERT INTO courses
+    (name, department_id)
+    VALUES
+    (?,?)
+  SQL
+  DB[:conn].execute(sql, attribute_values)
+
+  @id = DB[:conn].execute("SELECT last_insert_rowid() FROM courses")[0][0]
+end
+```
+Now when we run `rspec`, we see `undefined method new_from_db`. This method is going to take the information from the database and create a new instance by utilizing a method called `tap`. The `tap` method "yields `self` to the block, and then returns `self`. The primary purpose of this method is to “tap into” a method chain, in order to perform operations on intermediate results within the chain." In our case we are tapping into Course and creating a new instance while assigning it's attributes to specific rows.
+
+```ruby
+def self.new_from_db(row)
+  self.new.tap do |s|
+    s.id = row[0]
+    s.name =  row[1]
+    s.department_id = row[2]
+  end
+end
+```
+
+Now we can create a Courses table, insert a course and drop a course. Being able to find a course by it's name would also be useful, which is our next error. 
+
+We need to build a method called `find_by_name`. Again we can use `lib/student.rb` file as an example, replacing the necessary variables. Here we are selecting all from our courses table where the name is equal to the param we are passing in, then limiting it to one.
+
+```ruby
+ def self.find_by_name(name)
+    sql = <<-SQL
+      SELECT *
+      FROM courses
+      WHERE name = ?
+      LIMIT 1
+    SQL
+    result = DB[:conn].execute(sql,name)
+    result.map do |row| 
+      self.new_from_db(row)
+    end.first
+  end
+```
+Let's run `rspec` for our next error, `undefined method find_all_by_department_id`. Here we want to be able to find all courses by their `department_id`. The `SELECT` statement for this reads pretty clearly, `SELECT *` (`*` stands for all) from the courses table, where the `deparment_id` of the course equals the `id` we are passing in to the method, very simlilar to finding by name. Compare this code with your Student class to see the differences.
+
+```ruby
+def self.find_all_by_department_id(id)
+  sql = "SELECT * FROM courses WHERE department_id = ?"
+  result = DB[:conn].execute(sql,id)
+  result.collect do |row| 
+    self.new_from_db(row)
+  end
+end
+```
+Now we can create a Courses table, insert a course into the table, delete the table, find a course by name or department_id. What if we want to change the title of a course? We need a way to update the table.
+
+Run `rspec` and you will see `undefined method update`. Let's take a look at our `student.rb` file and see if there is anything to guide us. Looks like there is an `update` method built already, let's deconstruct it.
+
+The `update` method will take the values you are changing, find the object by it's id and update the instance. You can see our helper method `attribute_values` being passed in to the `DB[:conn].execute(sql, attribute_values, id)` part of the method.
```

Comment:
Maybe *this* is the point where you refactor the attribute_values out to a separate method!